### PR TITLE
Implement EIP-4844

### PR DESCRIPTION
**Pull request by roberto-bayardo**
this is a not-for-merging proof of concept implementation being used in the 4844 devnets.

### CODE REVIEW

File: consensus/misc/eip4844.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	targetGas := big.NewInt(params.TargetDataGasPerBlock)
+	if excessDataGas.Cmp(targetGas) < 0 {
+		return new(big.Int)
+	}
+	return new(big.Int).Set(excessDataGas.Sub(excessDataGas, targetGas))
+}
+
+// FakeExponential approximates factor * e ** (num / denom) using a taylor expansion
+// as described in the EIP-4844 spec.
+func FakeExponential(factor, num, denom *big.Int) *big.Int {
```
Comment (MariusVanDerWijden) [inline consensus/misc/eip4844.go:46]: This does not need to be exported right? It panics on {0,0,0} with div by zero. 
No way to trigger via GetDataGasPrice though

File: consensus/beacon/consensus.go [hunk 2]
```diff
@@ -256,8 +256,27 @@ func (beacon *Beacon) verifyHeader(chain consensus.ChainHeaderReader, header, pa
...
+	}
+	shanghai := chain.Config().IsShanghai(header.TimeBig())
+	if shanghai && header.WithdrawalsHash == nil {
+		return fmt.Errorf("missing withdrawalsHash")
+	}
+	// Verify existence / non-existence of withdrawalsHash.
+	if !shanghai && header.WithdrawalsHash != nil {
+		return fmt.Errorf("invalid withdrawalsHash: have %s, expected nil", header.WithdrawalsHash)
+	}
+	if chain.Config().IsSharding(header.TimeBig()) {
```
Comment (MariusVanDerWijden) [inline consensus/beacon/consensus.go:273]: I think we don't check that excessdataGas is not set before sharding
``` 
if !chain.Config().IsSharding(header.TimeBig()) {
		if header.ExcessDataGas != nil {
			return fmt.Errorf("invalid excessDataGas before fork: have %v, want <nil>", header.ExcessDataGas)
		}
	} else if err := misc.VerifyEip4844Header(chain.Config(), parent, header); err != nil {
		// Verify the header's EIP-4844 attributes.
		return err
	}
```

**Comment by MariusVanDerWijden** [comment]
QQ: I see that the `ExcessDataGas` is calculated during the finalization of the header, but I don't see where it is verified when receiving a block. Could you point me to where this is validated?

File: consensus/misc/eip4844.go (hunk 1 above)
Comment (roberto-bayardo) [reply consensus/misc/eip4844.go]: correct, changed to non-exported. re: the panic, should I have it return error or return 0 should someone pass in a 0 denom?

File: consensus/beacon/consensus.go (hunk 2 above)
Comment (roberto-bayardo) [reply consensus/beacon/consensus.go:273]: good catch, fixed

File: consensus/beacon/consensus.go [hunk 3]
```diff
@@ -316,29 +335,42 @@ func (beacon *Beacon) Prepare(chain consensus.ChainHeaderReader, header *types.H
...
 	}
+	// If withdrawals have been activated, process each one.
+	if chain.Config().IsShanghai(header.TimeBig()) {
+		for _, w := range withdrawals {
+			state.AddBalance(w.Address, w.Amount)
+		}
+	}
 	// The block reward is no longer handled here. It's done by the
 	// external consensus engine.
 	header.Root = state.IntermediateRoot(true)
```
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply consensus/beacon/consensus.go:355]: done

**Comment by roberto-bayardo** [comment]
> QQ: I see that the `ExcessDataGas` is calculated during the finalization of the header, but I don't see where it is verified when receiving a block. Could you point me to where this is validated?

it's not, will add.

**Comment by holiman** [comment]
This PR has been superseded by several other PRs, and will not be merged. I am closing this now, since it's no longer relevant for this repo.

**Comment by MariusVanDerWijden** [comment]
Thanks for your work @roberto-bayardo! It was a great help when I implemented our 4844 version. I think I even lifted some of your test cases

**Pull request closed without merging**